### PR TITLE
Add wetting and drying capability to local time-stepping scheme

### PR DESCRIPTION
**Pull request by gcapodag**
This PR adds a wetting and drying capability to the local time-stepping (LTS) scheme for the barotropic ocean already available in master. The PR also adds several improvements for the LTS scheme not related to wetting and drying like for instance the `upwind` option for the `layerThickEdgeFlux` and inline tidal boundary conditions needed to run the drying slope test case from Compass. A separate PR on Compass will include drying slope, dam break, and parabolic bowl test cases for LTS.

This PR is bit-for-bit for E3SM mainline simulations, which do not include LTS or wetting and drying.

[BFB]

### CODE REVIEW

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (cbegeman) [inline components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:392]:
```suggestion
    if (config_use_wetting_drying) then
```

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (cbegeman) [inline components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:1339]:
```suggestion
      call mpas_log_write('ERROR: config_thickness_flux_type selected not implemented for LTS', messageType=MPAS_LOG_CRIT)
```

File: components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F
Comment (cbegeman) [inline components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F:406]:
```suggestion
   subroutine ocn_compute_wetting_velocity_factor(wettingVelocityFactor, layerThicknessTmp, normalVelocity, iCell, k)!{{{
```
I think it's better to leave off `cell`. It could be confusing since `wettingVelocityFactor` is an edge quantity.

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (cbegeman) [inline components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:502]: I'm confused about why you need this factor here. Have you scaled each of the terms here already? It seems that you are double-scaling at least a few of them.

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (cbegeman) [inline components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:737]: Same here, I would think you only need to scale the last term by `1-wettingVelocityFactor`

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (cbegeman) [inline components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:716]: If `normalVelocityTend` comes from `ocn_tend_vel` then it has already been scaled by `(1.0_RKIND - wettingVelocityFactor(:, iEdge))`, but I'm a bit confused by what `normalVelocityTendSum1st` refers to since the first term does not appear to be a velocity tendency.

**Review by cbegeman — CHANGES_REQUESTED** [review]
@gcapodag Exciting work!

I posted a few questions here because I found it hard to keep track of where you're scaling `normalVelocityTend` terms by `wettingVelocityFactor` and where you're scaling `normalVelocity` terms. In previous implementations of wetting and drying, we scale each of these only once (here, I guess it would be once per stage). Can you clarify for me how this is working here or point me to a design doc?

I haven't done any testing yet. Let me know when the compass LTS PR is ready.

**Comment by mark-petersen** [comment]
@gcapodag now that the AB2 PR is merged (#5989), it is a good time to rebase these commits to the head of master. Since you merged master into this branch a few times along the way, I recommend starting with master and cherry-picking your commits. Thanks!

**Comment by xylar** [comment]
> I recommend starting with master and cherry-picking your commits.

@mark-petersen, an interactive rebase in which you pick the commits you want and remove the rest from the list is an easier way to cherry-pick several commits.

File: components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F
Comment (gcapodag) [reply components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F:406]: I changed it to: `ocn_wetting_velocity_factor_on_cell_edges`

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (gcapodag) [reply components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:502]: Thanks for your feedback @cbegeman . There is no double scaling, except the one coming from previous applications of the limiter at earlier stages if that was needed. `wettingVelocityFactor` is computed in `ocn_lts_wd` and then used once in the `flux` for the `h` tendency (in the call to `ocn_tend_lts`) and then subsequently where we advance the normal velocity solution. In LTS we have ad-hoc routines to compute the tendencies (`ocn_tend_lts`) due to the splitting, so `ocn_tend_vel` is not used. I think that a double scaling could be happening though for RK4, because [here](https://github.com/E3SM-Project/E3SM/blob/839d1cf38fe80aa59a934f3ec5a2c260f3b8aa72/components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F#L281) the old velocity is limited and then again multiplied by the factor [here](https://github.com/E3SM-Project/E3SM/blob/839d1cf38fe80aa59a934f3ec5a2c260f3b8aa72/components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_rk4.F#L1249). The tendency is also limited twice, where the provis solution is advanced and in the [call](https://github.com/E3SM-Project/E3SM/blob/839d1cf38fe80aa59a934f3ec5a2c260f3b8aa72/components/mpas-ocean/src/shared/mpas_ocn_tendency.F#L469) to `ocn_tend_vel`, as you mentioned.

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (gcapodag) [reply components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:737]: if we do that then the velocity would not be set to zero since the old velocity is not limited

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (gcapodag) [reply components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:716]: or, I should say, it is used only for the slow terms

**Comment by gcapodag** [comment]
Thank you for your input @cbegeman , I replied to your comments, hopefully my replies make sense. As soon as https://github.com/MPAS-Dev/compass/pull/716 goes in, I will submit the compass PR as well.

**Comment by gcapodag** [comment]
@mark-petersen @xylar please let me know if the "rebase" has gone through. thanks!

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (cbegeman) [reply components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:716]: @gcapodag Thanks for explaining. This makes sense to me.

**Comment by gcapodag** [comment]
@cbegeman I might get the compass PR going so it will be possible to test this, and when the init mode changes to the drying slope will be merged, I will add those as well. How does this sound?

**Comment by cbegeman** [comment]
> @cbegeman I might get the compass PR going so it will be possible to test this, and when the init mode changes to the drying slope will be merged, I will add those as well. How does this sound?

Yes, that sounds great. Once you post the PR then I will review it along with this PR. Merging the compass PR after the E3SM PR is merged is the way to go. Thanks!

**Comment by gcapodag** [comment]
Testing for this PR can be done from this Compass PR: https://github.com/MPAS-Dev/compass/pull/738

**Comment by mark-petersen** [comment]
Compiled with gnu on chicoma and intel in chrysalis. With the later, results are bfb against master branch point using nightly compass suite with default flags. That is a sanity test, as I'm not running LTS. But so far so good...

**Comment by rljacob** [comment]
Needs an approval.

**Comment by mark-petersen** [comment]
Since I last looked at this Dec 5, I started with this branch and merged in 09491ec080, which is master from Dec 21 just before the stand-alone bug in https://github.com/E3SM-Project/E3SM/issues/6133. I then recompiled stand-alone with intel on chrysalis, and also ran the E3SM tests:
```
cd cime/scripts
./create_test SMS_Ld62.T62_oQU240.GMPAS-IAF.chrysalis_intel -q acme-small --walltime 1:00:00
./create_test SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.chrysalis_gnu -q acme-small --walltime 30:00
```
These both pass the compile and smoke test.

**Comment by gcapodag** [comment]
@mark-petersen sounds good, thanks!

File: components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F
Comment (cbegeman) [reply components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:502]: Thank you for this detailed explanation! That helped me understand what was going on.

Comment (gcapodag) [reply components/mpas-ocean/src/mode_forward/mpas_ocn_time_integration_lts.F:502]: Of course, thanks for reviewing!

**Comment by gcapodag** [comment]
@cbegeman the compass PR should be ready for you try out with this one as well

**Comment by gcapodag** [comment]
@cbegeman does this PR have a green light on your end?

**Comment by cbegeman** [comment]
@gcapodag Thanks for reminding me to approve this. I'll do it now, given that my compass tests were successful. Great work!

**Review by cbegeman — APPROVED** [review]

**Comment by gcapodag** [comment]
@cbegeman thank you!

**Comment by mark-petersen** [comment]
Tested current head and rebased locally onto master. Compiled with intel on chrysalis and gnu on chicoma, with both debug and optimized. Ran nightly compass test suite with all (that does not test this feature, of course). Both gnu and intel are bfb on all tests against master branch point. Also tested and passed E3SM
```
SMS_Ln9.T62_oQU240.GMPAS-IAF.chrysalis_gnu
SMS_Ln9.T62_oQU240.GMPAS-IAF.chrysalis_intel
```

**Review by mark-petersen — APPROVED** [review]
I tested with the associated compass test cases at https://github.com/MPAS-Dev/compass/pull/738#issuecomment-1915743613. Approving based on the beautiful results in these test cases, and that there is no alterations of the mainline model results for this stealth feature.

**Comment by gcapodag** [comment]
Thank you very much @mark-petersen!

**Comment by mark-petersen** [comment]
FYI for the E3SM review, the Local Time Stepping feature that is altered here is only used in MPAS-Ocean standalone for single layer simulations right now.

**Comment by jonbob** [comment]
Passes:
* ERS.ne30pg2_r05_EC30to60E2r2.GPMPAS-JRA.chrysalis_intel.mosart-rof_ocn_2way
* ERS_Ld5.T62_oQU120.CMPASO-NYF.chrysalis_intel
* SMS_D_Ld3.T62_oQU120.CMPASO-IAF.chrysalis_intel

merged to next

**Comment by jonbob** [comment]
merged to master

**Pull request merged**
